azurerm/resource_arm_application_gateway.go

```diff
@@ -0,0 +1,1685 @@
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		name := data["name"].(string)
+		defaultBackendAddressPoolName := data["default_backend_address_pool_name"].(string)
+		defaultBackendAddressPoolID := fmt.Sprintf("%s/backendAddressPools/%s", gatewayID, defaultBackendAddressPoolName)
+		defaultBackendHTTPSettingsName := data["default_backend_http_settings_name"].(string)
+		defaultBackendHTTPSettingsID := fmt.Sprintf("%s/backendHttpSettingsCollection/%s", gatewayID, defaultBackendHTTPSettingsName)
+
+		pathRules := []network.ApplicationGatewayPathRule{}
+		for _, ruleConfig := range data["path_rule"].([]interface{}) {
+			ruleConfigMap := ruleConfig.(map[string]interface{})
+
+			ruleName := ruleConfigMap["name"].(string)
+
+			rulePaths := []string{}
+			for _, rulePath := range ruleConfigMap["paths"].([]interface{}) {
+				rulePaths = append(rulePaths, rulePath.(string))
+			}
+
+			rule := network.ApplicationGatewayPathRule{
+				Name: &ruleName,
+				ApplicationGatewayPathRulePropertiesFormat: &network.ApplicationGatewayPathRulePropertiesFormat{
+					Paths: &rulePaths,
+				},
+			}
+
+			if backendAddressPoolName := ruleConfigMap["backend_address_pool_name"].(string); backendAddressPoolName != "" {
+				backendAddressPoolID := fmt.Sprintf("%s/backendAddressPools/%s", gatewayID, backendAddressPoolName)
+				rule.ApplicationGatewayPathRulePropertiesFormat.BackendAddressPool = &network.SubResource{
+					ID: &backendAddressPoolID,
+				}
+			}
+
+			if backendHTTPSettingsName := ruleConfigMap["backend_http_settings_name"].(string); backendHTTPSettingsName != "" {
+				backendHTTPSettingsID := fmt.Sprintf("%s/backendHttpSettingsCollection/%s", gatewayID, backendHTTPSettingsName)
+				rule.ApplicationGatewayPathRulePropertiesFormat.BackendHTTPSettings = &network.SubResource{
+					ID: &backendHTTPSettingsID,
+				}
+			}
+
+			pathRules = append(pathRules, rule)
+		}
+
+		pathMap := network.ApplicationGatewayURLPathMap{
+			Name: &name,
+			ApplicationGatewayURLPathMapPropertiesFormat: &network.ApplicationGatewayURLPathMapPropertiesFormat{
+				DefaultBackendAddressPool: &network.SubResource{
+					ID: &defaultBackendAddressPoolID,
+				},
+				DefaultBackendHTTPSettings: &network.SubResource{
+					ID: &defaultBackendHTTPSettingsID,
+				},
+				PathRules: &pathRules,
+			},
+		}
+
+		pathMaps = append(pathMaps, pathMap)
+	}
+
+	return &pathMaps
+}
+
+func expandApplicationGatewayAuthenticationCertificates(d *schema.ResourceData) *[]network.ApplicationGatewayAuthenticationCertificate {
+	configs := d.Get("authentication_certificate").([]interface{})
+	authCerts := make([]network.ApplicationGatewayAuthenticationCertificate, 0, len(configs))
+
+	for _, configRaw := range configs {
+		raw := configRaw.(map[string]interface{})
+
+		name := raw["name"].(string)
+		data := raw["data"].(string)
+
+		// data must be base64 encoded
+		data = base64.StdEncoding.EncodeToString([]byte(data))
```

Comment:
in other places this is the case (e.g. customData in VM's), we only base64 encode if it's not already encoded - could we update this to match?